typing-practice-react/src/index.js

```diff
@@ -0,0 +1,179 @@
+    let charArray = text.innerText;
+    const userInput = event.key;
+    const userInputIsCorrect = charArray[state.charIndex] === userInput;
+    const isFirstTry = state.tryCounter == 0;
+    const isBeginOfTyping = state.userKeyTypeCount == 0;
+    
+    if (isBeginOfTyping) {
+      this.beginCounting(state, statsTextForSeconds, todayStatsText, totalStatsText, lastSetStatsText);
+    }
+
+    if (state.generateTextButtonIsClicked) {
+      keyboardKeysArray.forEach(function(keyBoardKeyEntryInArray) {
+        const isTheCorrectlyTypedKey = keyBoardKeyEntryInArray[0] == (event.key).toUpperCase();
+        if (isFirstTry && userInputIsCorrect) {
+          if (isTheCorrectlyTypedKey) {
+            keyBoardKeyEntryInArray[1]++; 
+          }
+        }
+  
+        if (isTheCorrectlyTypedKey) {
+          keyBoardKeyEntryInArray[2]++;
+        }
+      });
+      
+      state.extendedStatsString = "";
+      keyboardKeysArray.forEach(function(keyBoardKeyEntryInArray) {
+        let computedValue = Math.round(keyBoardKeyEntryInArray[1] / keyBoardKeyEntryInArray[2] * 100);
+        if (keyBoardKeyEntryInArray[1] == 0) {
+          computedValue = 0;
+        }
+        
+        state.extendedStatsString += `${keyBoardKeyEntryInArray[0]} ${computedValue}%<br>`;
+      });
+      
+      extendedStatsText.innerHTML = state.extendedStatsString;
+  
+      if (!userInputIsCorrect) {
+        state.tryCounter ++;
+        state.userMistakesCount ++;
+        localStorage.todayMistypes = parseInt(localStorage.todayMistypes) + 1;
+        localStorage.totalMistypes = parseInt(localStorage.totalMistypes) + 1;
+      }
+      else {
+        state.tryCounter = 0;
+      }
+  
+      if (userInput != 'Space') {
+        state.userKeyTypeCount ++;
+        localStorage.todayCharsTyped = parseInt(localStorage.todayCharsTyped) + 1;
+        localStorage.totalCharsTyped = parseInt(localStorage.totalCharsTyped) + 1;
+      }
+      const endOfArrayIsReached = state.charIndex === charArray.length - 1;
+      if (endOfArrayIsReached) {
+        clearInterval(this.incrementSecondsInterval);
+        charArray = await generateFinalText(text);
+        resetProgress(state, text, this.lastSetStatsText.current, this.todayStatsText.current);
+        localStorage.todayAmountOfSets = parseInt(localStorage.todayAmountOfSets) + 1;
+        localStorage.totalAmountOfSets = parseInt(localStorage.totalAmountOfSets) + 1;
+  
+        return;
+      }  
+  
+      state.charIndex = handleKeyDownEvent(event, text, lastSetStatsText, todayStatsText, charArray, state.charIndex, state.seconds, state.userKeyTypeCount, state.userMistakesCount, userInputIsCorrect);
+    }
+}
+  render() {
+    const statsText = displayStats(this.state.userMistakesCount, this.state.userKeyTypeCount, this.state.seconds);
+    const todayStatsText = displayTodayStats();
+    const totalStatsText = displayTotalStats();
+    const secondsStatsText = `Seconds ${this.state.seconds}`;
+
+    return (
```

Comment:
I get what you meant in the morning today. The render function here seems really bloated and doesn't align with the ideas of the react framework.
You can try to move the stats on its own component for example.
In the projects later on I wouldn't allow this approach.

Reply:
Yes, you are totally right, and In future I will do a different approach but for this project, I guess I will leave it as it is.